pwnlib/fmtstr.py

```diff
@@ -151,33 +174,36 @@ def fmtstr_payload(offset, writes, numbwritten=0, write_size='byte'):
     if write_size not in ['byte', 'short', 'int']:
         log.error("write_size must be 'byte', 'short' or 'int'")
 
+    # number: number of write_size in a word
+    # step: amount of bytes in write_size
+    # mask: mask for written values
+    # formatz: format size modifier for %n
+    # declage: amount of bits in write_size
     number, step, mask, formatz, decalage = config[context.bits][write_size]
 
-    # add wheres
-    payload = ""
-    for where, what in writes.items():
+    # create packed addrs using wheres
+    packed_addrs = ""
+    for where, _ in writes.items():
         for i in range(0, number*step, step):
-            payload += pack(where+i)
-
-    numbwritten += len(payload)
-    fmtCount = 0
-    for where, what in writes.items():
-        for i in range(0, number):
-            current = what & mask
-            if numbwritten & mask <= current:
-                to_add = current - (numbwritten & mask)
-            else:
-                to_add = (current | (mask+1)) - (numbwritten & mask)
-
-            if to_add != 0:
-                payload += "%{}c".format(to_add)
-            payload += "%{}${}n".format(offset + fmtCount, formatz)
-
-            numbwritten += to_add
-            what >>= decalage
-            fmtCount += 1
-
-    return payload
+            packed_addrs += pack(where+i)
+
+    # decide whether the packed addrs should be placed at the beginning or end
+    # of the payload
+    if '\x00' in packed_addrs:
+        # place at end of payload
+        pad_chr = ' '
+        pad_len = 0 # length to pad the format section to, in processor words
+        while True:
+            # try increasing amounts of padding
+            pad_len += 1
+            fmts = fmts_with_pad(numbwritten, pad_len)
+            if pad_len * context.bytes >= len(fmts):
+                break
+        return fmts.ljust(pad_len * context.bytes, pad_chr) + packed_addrs
```

Comment:
Tiny nit: I would prefer that we just update `packed_addrs` and have a single return statement at the end.

Reply:
Can we create a new variable `payload` for this? It might be confusing to have `packed_addrs` store the entire payload and not just the packed addresses.

Reply:
Either way, I mistakenly deleted the fork for this PR so I don't think I can push to it anymore.